include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+  }
+
+  /**
+   * Call on a stream context to indicate that the request metadata has arrived.
+   */
+  virtual ProxyAction onResponseMetadata() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  // Root Context async events
+
+  /**
+   * Called on a Root Context when a response arrives for an outstanding httpCall().
+   * @param token is the token returned by the corresponding httpCall().
+   * @param headers are the number of headers in the response.
+   * @param body_size is the size in bytes of the response body.
+   * @param trailers is the number of trailres in the response.
+   */
+  virtual void onHttpCallResponse(uint32_t /* token */, uint32_t /* headers */,
+                                  uint32_t /* body_size */, uint32_t /* trailers */) {}
+
+  /**
+   * Called on a Root Context to indicate that an Inter-VM shared queue message has arrived.
+   * @token is the token returned by registerSharedQueue().
+   */
+  virtual void onQueueReady(uint32_t /* token */) { unimplemented(); }
+
+  // Context
+
+  /**
+   * Call when a stream has completed (both sides have closed) or on a Root Context when the VM is
+   * shutting down.
+   * @return true for stream contexts or for Root Context(s) if the VM can shutdown, false for Root
+   * Context(s) if the VM should wait until the Root Context calls the proxy_done() ABI call.  Note:
+   * the VM may (optionally) shutdown after some configured timeout even if the Root Context does
+   * not call proxy_done().
+   */
+  virtual bool onDone();
+
+  /**
+   * Call when a stream should log the final status of the stream. Occurs after onDone().
+   */
+  virtual void onLog();
+
+  /**
+   * Call when no further stream calls will occur.  This will cause the corresponding Context in the
+   * VM to be deleted.
+   */
+  virtual void onDelete();
+
+  /**
+   * Will be called on sever Wasm errors. Callees may report and handle the error (e.g. via an
+   * Exception) to prevent the proxy from crashing.
+   */
+  virtual void error(string_view message) {
+    std::cerr << message << "\n";
+    abort();
+  }
+
+  /**
+   * Called by all functions which are not overriden with a proxy-specific implementation.
+   */
+  virtual void unimplemented() { error("unimplemented proxy-wasm API"); }
+
+  //
+  // Calls from the VM>
+  // Note: most of these return a WasmResult, assume:
+  // @return a WasmResult with the status of the call.
+  //
+  //
+
+  /**
+   * Log a message.  Note: log_prefix() is *not* auto-prepended end may be useful to provide some
+   * context.
+   */
+  virtual WasmResult log(uint64_t /* level */, string_view /* message */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Enables a periodic timer with the given period or sets the period of an existing timer. Note:
+   * the timer is associated with the Root Context of whatever Context this call was made on and
+   * there is only one timer available per Root Context.
```

Comment:
Having a single timer seems like a pretty big limitation. Why not have a timer id and an ability to allocate a timer? I'm pretty sure this is going to come up very quickly if it hasn't already?

Reply:
SIGALRM is a pretty clear counter-example.  However, we can do that.  Could you make the comment on the spec and get it changed there?  Then we can update this file to match the spec.

Reply:
How do I make a comment on the spec?

Reply:
Let let me talk to @PiotrSikora.  We need to create a reviewed branch.

Reply:
I don't see any timer functions in the ABI doc here? https://github.com/proxy-wasm/spec/pull/1/files

Reply:
`proxy_set_tick_period_milliseconds is used.  We could add a token_ptr and if the target token was zero we could use that to indicate that the system should allocate a new timer, but if it was non-zero then it would be the timer to set a new period for.  WDYT? I'll add this comment on the spec as well.

Reply:
My personal feeling is that the ABI should directly support timer allocation, since many/most host systems would support that. If a host only supports SIGALRM for example, it would be responsible for doing it's own scheduling which seems reasonable to me given such an uncommon case. So my advice would be to have the ABI directly support timer allocation in addition to a tick mechanism. If for whatever reason we don't want to do that I think what you propose is fine, though it would require implementing our own scheduling logic which to me seems not worth it.

Reply:
set_tick_period allocates a timer if the period is non-zero as per the documentation.  It deallocates the timer if the period is zero.   What I am proposing is that we have a way for the system to return a token representing the allocated timer and for the SDK to provide a token.  The on_tick call would also be augmented with a uint32_t timer_token

Reply:
Sure sgtm